### PR TITLE
Add android/{shell.nix,nix-global-sdk.sh,.envrc}

### PR DESCRIPTION
**Pull request by nicknovitski**
I've decided that I think we should do this, but: 
- I'm open to objections
- I haven't tested the built apk except to confirm that it runs on my phone
- It's blocked on nixos/nixpkgs#48848

## Why I bothered trying this out

When I set up the android CI build, I looked over [the nixpkgs tools for android build environments](https://github.com/NixOS/nixpkgs/tree/9a2fd03d67b5264de4f93206d2e39d92c4bbb942/pkgs/development/mobile/androidenv), and decided to instead use a set of docker images published by circle (and which they label as "alpha").

Recently, @dsokal asked me if nix could be leveraged for the purpose of installing the android sdk and ndk to developer machines.  Looking over the packaging again, to explain precisely why I had decided not to use it, I believe I figured out how to fix my objections.

## What it gives us

In general, controlling dependencies across machines, making it easier for developers to create and update working build environments, and to prevent and reproduce CI failures, which in turn makes it easier to accept contributions.

More specifically, from the android directory, the command `nix-shell` will provide the same environment on any computer (minus OS differences), providing everything needed for gradle builds, downloading anything it needs to.  The environment and any code depending on it can change in the same commit.

In the same directory, the script `nix-global-sdk.sh` creates or updates `~/Library/Android/sdk-nix` to contain the same sdk and ndk, which can be used by Android Studio.

(Note that Android Studio refuses to download any version of the NDK besides the latest.  So this gives us control over that for the first time.)

## What it costs us

I believe not too many people use the nixpkgs android infrastructure.  The upstream PR this one requires includes the following changes
- add a config option for the sdk package for users to indicate they accept the eula
- patch some paths in binaries in the ndk package
- patch some paths in scripts in the build-tools package
- add an entire infrastructure for packaging all build-tools versions and specifying them as arguments to the sdk-package-producing function was necessary (I promise I tried other approaches first)
- assorted package updates

We would not need to do all of this every time we wanted to add or change a dependency (it's especially unlikely that we'd be blocked on an upstream PR, most changes would work with just local overrides and packages in this repository), but I think it gives an idea: changes of that sort will be linked to packaging changes, and if no one else has contributed it, we'd have to write it ourselves.  For example, ndk r18b is currently not packaged, and would need a patch file to work.

### CODE REVIEW

File: android/shell.nix
Comment (ide) [inline android/shell.nix]:
```suggestion
     abiVersions = [ "x86" "x86_64" ];
```

File: default.nix
Comment (ide) [inline default.nix]: Can you move this fork to the expo org?

**Comment by ide** [comment]
> In the same directory, the script nix-global-sdk.sh creates or updates ~/Library/Android/sdk-nix to contain the same sdk and ndk, which can be used by Android Studio.

Can the downloaded SDK/NDK also be used for builds from the command line (`android/run.sh`)? I think there is something nice about being able to do the builds from the host environment, as opposed to the Nix shell, and where the Nix shell shines is in CI, especially if we wanted to look into GitHub Actions or other CI providers.

**Review by ide — APPROVED** [review]

**Comment by nicknovitski** [comment]
`nix-shell --run ./run.sh` just works.  I actually hadn't even looked at the script until now, so there's a nice natural experiment.

The nix shell is just a bash subshell with some environment variables set, most notably PATH; unlike docker, everything is run on the host.  If you mean, it's nice to have `adb` in PATH even without explicitly running nix-shell, then I agree.  We don't have any automatic setup for that currently, right?  I could add an `android/.envrc` that set ANDROID_SDK_ROOT to ~/Library/Android/sdk-nix, and added the right subdirectories to the path, and even detected and complained if the link was out of date.

File: default.nix
Comment (nicknovitski) [reply default.nix]: I could, but I don't want to merge a change that points our pin at a fork.  That would effectively mean maintaining our own nixpkgs channel, which I do not want.  Instead I'm going to only merge this PR when the upstream changes it depends on get merged.

Comment (nicknovitski) [reply default.nix]: Phew, they have been merged.

**Comment by nicknovitski** [comment]
I added an .envrc that sets the _ROOT and _HOME variables to ~/Library/Android/sdk-nix, and adds adb and similar tools to PATH.

I tried having it check if either link is out of date, but I did not like the performance impact.

Before:
```
$ time bash -cl ""

real    0m0.439s
user    0m0.300s
sys     0m0.151s
```
After:
```
$ time bash -cl ""

real    0m0.885s
user    0m0.363s
sys     0m0.265s
```
That runs each time direnv loads, so each time you enter the directory or open a new shell in it, including tmux windows.  It was very noticeable, even with my already-slow shell.

**Comment by nicknovitski** [comment]
The necessary upstream changes have been merged, but I'll wait until they make it through a release channel.

**Pull request merged**
